### PR TITLE
ID Token validation

### PR DESCRIPTION
**Pull request by arvindsinghtomar**

### CODE REVIEW

**Comment by nynymike** [comment]
Hi guys. Please let Arvind know if there is anything we can do to get this right.

File: library/java/net/openid/appauth/Utils.java [hunk 1]
```diff
@@ -55,4 +58,17 @@ public static void closeQuietly(InputStream in) {
...
         }
     }
+
+    /**
+     * Decodes Base 64 url.
+     */
+    public static String decodeBase64urlNoPadding(String strEncoded)
+        throws UnsupportedEncodingException {
+        if (strEncoded != null) {
+            byte[] decodedBytes = Base64.decode(strEncoded, Base64.NO_PADDING);
```
Comment (WilliamDenniss) [inline library/java/net/openid/appauth/Utils.java]: I'm pretty sure you need `URL_SAFE` too.

**Comment by iainmcgin** [comment]
Just so you know I'm not ignoring this: I'll try and review thoroughly on Monday.

File: library/java/net/openid/appauth/AuthorizationService.java [hunk 2]
```diff
@@ -559,4 +577,120 @@ protected void onPostExecute(JSONObject json) {
...
+        @Override
+        protected JSONObject doInBackground(Void... voids) {
+            InputStream is = null;
+            try {
+                Uri uri = (mResponse.request.configuration.discoveryDoc.getJwksUri());
+                HttpURLConnection conn =
+                        mClientConfiguration.getConnectionBuilder()
+                        .openConnection(uri);
+                conn.setRequestMethod("GET");
+                //conn.setDoOutput(true);
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/AuthorizationService.java]: if these commented out code sections are not needed, just remove them.

File: library/java/net/openid/appauth/AuthorizationService.java [hunk 3]
```diff
@@ -559,4 +577,120 @@ protected void onPostExecute(JSONObject json) {
...
+
+                is = new BufferedInputStream(conn.getInputStream());
+                // readStream(in);
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/AuthorizationService.java]: Same as above: remove if not needed.

File: library/java/net/openid/appauth/AuthorizationService.java [hunk 4]
```diff
@@ -559,4 +577,120 @@ protected void onPostExecute(JSONObject json) {
...
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(JSONObject json) {
+            JSONObject requiredJson = null;
+
+            try {
+                String[] split = mResponse.idToken.split("\\.");
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/AuthorizationService.java]: Validate that there are exactly three sections before proceeding.

File: library/java/net/openid/appauth/AuthorizationService.java [hunk 5]
```diff
@@ -559,4 +577,120 @@ protected void onPostExecute(JSONObject json) {
...
+                if (mException != null
+                        || requiredJson == null
+                        || !json.optString("error", "").equals("")
+                        || !jsonObjectHeader.getString("kid").equals(requiredJson.getString("kid"))
+                        || !jsonObjectBody.getString("aud").equals(mResponse.request.clientId)) {
+                    mCallback.onTokenValidationRequestCompleted(false, mException);
+                    return;
+                }
+
+                Calendar calendar = Calendar.getInstance();
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/AuthorizationService.java]: I'm also not convinced that you're gaining much from calendar beyond directly using the exp value. Code can probably just be:

```java
long expirationTime = Long.parseLong(jsonObjectBody.getString("exp") * 1000L;
if (expirationTime > mClock.getCurrentTimeMillis()) { ... } else { ... }
```

**Comment by arvindsinghtomar** [comment]
@iainmcgin I have done change you suggested in comment

**Comment by iainmcgin** [comment]
Looks like you still have checkstyle warnings - please run `./gradlew check`and ensure that this builds successfully. Also, you could separate your README change for Gluu into a separate pull request, as it is generally unrelated to the rest of this change.

**Comment by ravibhojwani** [comment]
Aeternity are working some magic

**Comment by arvindsinghtomar** [comment]
@iainmcgin Now I have updated with the latest code, completed ./gradlew check and removed readme-gluu.md for now. Please review it.

**Comment by arvindsinghtomar** [comment]
@iainmcgin Now I have updated with the latest code, completed ./gradlew check and removed readme-gluu.md for now. Please review it.

**Comment by arvindsinghtomar** [comment]
@iainmcgin please review my PR

**Comment by iainmcgin** [comment]
Apologies, I've not had time to take a look yet.

**Comment by nynymike** [comment]
Thanks Iain! We really want to publish the Gluu README!

**Comment by iainmcgin** [comment]
Closing as #385 provides our preferred approach to ID token structure validation, without the signature verification. We have decided against signature verification within the core library as it isn't required for tokens acquired through an authorization code flow.

**Comment by nynymike** [comment]
Nice! That makes sense, it's easier to see if the TLS certificate `cn` matches the `iss`. You can still validate the signature, which might be good for higher assurance use cases.

**Pull request closed without merging**